collections/binary_search_tree.ts

```diff
@@ -255,13 +252,13 @@ export class BinarySearchTree<T> implements Iterable<T> {
           replacementNode;
       }
 
-      if (successorNode !== removeNode) {
-        removeNode.value = successorNode.value;
-        removeNode = successorNode;
+      if (successorNode !== node) {
+        node.value = successorNode.value;
+        node = successorNode;
       }
       this._size--;
     }
-    return removeNode;
+    return node;
```

Comment:
From my previous comment, I would change this to `return successorNode !== node ? successorNode : null;`

I'd also suggest adding a documentation comment for this function indicating it will remove the node from the tree and return the successor node if there is one.